File: Assets/src/model/tasks/MoveRandomlyTask.cs
using System;
using System.Collections.Generic;
using System.Linq;
using UnityEngine;

[System.Serializable]
class MoveRandomlyTask : DoOnceTask {
  private Func<Tile, bool>? predicate;

  public MoveRandomlyTask(Actor actor, Func<Tile, bool>? predicate = null) : base(actor) {
    this.predicate = predicate;
  }

  protected override BaseAction GetNextActionImpl() {
    return GetRandomMove(actor, predicate);
  }

  public static BaseAction GetRandomMove(Actor actor, Func<Tile, bool>? predicate = null) {
    var adjacentTiles = actor.floor?.GetAdjacentTiles(actor.pos).Where((tile) => tile.CanBeOccupied());
    if (predicate != null) {
      adjacentTiles = adjacentTiles.Where(predicate);
    }
    if (adjacentTiles?.Any() ?? false) {
      Vector2Int pos = Util.RandomPick(adjacentTiles).pos;
      return new MoveBaseAction(actor, pos);
    } else {
      return new WaitBaseAction(actor);
    }
  }
}

File: web/src/model/tasks/MoveRandomlyTask.ts
import { DoOnceTask } from '../ActorTask';
import { MoveBaseAction, WaitBaseAction, type BaseAction } from '../BaseAction';
import { MyRandom } from '../../core/MyRandom';
import type { Actor } from '../Actor';
import type { Tile } from '../Tile';

export class MoveRandomlyTask extends DoOnceTask {
  private predicate: ((t: Tile) => boolean) | null;

  constructor(actor: Actor, predicate: ((t: Tile) => boolean) | null = null) {
    super(actor);
    this.predicate = predicate;
  }

  protected getNextActionImpl(): BaseAction {
    return MoveRandomlyTask.getRandomMove(this.actor, this.predicate);
  }

  static getRandomMove(actor: Actor, predicate: ((t: Tile) => boolean) | null = null): BaseAction {
    let tiles = actor.floor?.getAdjacentTiles(actor.pos).filter(t => t.canBeOccupiedBy(actor)) ?? [];
    if (predicate) {
      tiles = tiles.filter(predicate);
    }
    if (tiles.length > 0) {
      return new MoveBaseAction(actor, MyRandom.Pick(tiles).pos);
    }
    return new WaitBaseAction(actor);
  }
}
